rfcs/20210510-binary-size-reduction/ISA_extension.md

```diff
@@ -0,0 +1,117 @@
+implementation logic straight.
+
+## Implementation Details
+Build system will define the following set of macros:
+
+```cpp
+...
+// Primitives ISA controls
+#cmakedefine01 BUILD_PRIMITIVE_ISA_ALL
+#cmakedefine01 BUILD_AVX512
+#cmakedefine01 BUILD_AVX2
+#cmakedefine01 BUILD_SSE41
+...
+```
+
+Based on this `BUILD_XXX` macros, the following internal macros would be set:
+
+```cpp
+// src/common/impl_registration.hpp:
+
+#if BUILD_PRIMITIVE_ISA_ALL
+#define REG_ISA_AMX(...) __VA_ARGS__
+#define REG_ISA_AVX512(...) __VA_ARGS__
+#define REG_ISA_AVX2(...) __VA_ARGS__
+#define REG_ISA_SSE41(...) __VA_ARGS__
+#elif BUILD_AVX512
+#define REG_ISA_AMX(...)
+#define REG_ISA_AVX512(...) __VA_ARGS__
+#define REG_ISA_AVX2(...) __VA_ARGS__
+#define REG_ISA_SSE41(...) __VA_ARGS__
+#elif BUILD_AVX2
+#define REG_ISA_AMX(...)
+#define REG_ISA_AVX512(...)
+#define REG_ISA_AVX2(...) __VA_ARGS__
+#define REG_ISA_SSE41(...) __VA_ARGS__
+#elif BUILD_SSE41
+#define REG_ISA_AMX(...)
+#define REG_ISA_AVX512(...)
+#define REG_ISA_AVX2(...)
+#define REG_ISA_SSE41(...) __VA_ARGS__
+#endif
+```
+
+Based on this set of macros, `CPU_INSTANCE_X64` would be extended with the
+following list:
+```cpp
+// src/cpu/cpu_engine.hpp:
+
+#define CPU_INSTANCE_X64_AVX2(...) REG_ISA_AVX2(DNNL_X64_ONLY(CPU_INSTANCE(__VA_ARGS__)))
+#define CPU_INSTANCE_X64_AVX512(...) REG_ISA_AVX512(DNNL_X64_ONLY(CPU_INSTANCE(__VA_ARGS__)))
+#define CPU_INSTANCE_X64_AMX(...) REG_ISA_AMX(DNNL_X64_ONLY(CPU_INSTANCE(__VA_ARGS__)))
+```
+
+Finally, these internal macros are applied inside primitive implementation
+lists:
+
+```cpp
+// src/cpu/cpu_convolution_list.cpp:
+const std::map<conv_impl_key_t, std::vector<pd_create_f>> impl_list_map {
+    {{forward, s8, s8, f32}, {
+        REG_CONV_P_FWD(REG_IP_P_FWD(CPU_INSTANCE_X64(ip_convolution_fwd_t)))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64_AMX(jit_avx512_core_amx_1x1_convolution_fwd_t))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64_AVX512(jit_avx512_core_amx_convolution_fwd_t))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64_AVX512(jit_avx512_core_x8s8s32x_1x1_convolution_fwd_t))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64_AVX512(jit_avx512_core_x8s8s32x_convolution_fwd_t))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64_AVX2(jit_uni_x8s8s32x_1x1_convolution_fwd_t<avx2>))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64_AVX2(jit_uni_x8s8s32x_convolution_fwd_t<avx2>))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64(jit_uni_x8s8s32x_1x1_convolution_fwd_t<sse41>))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64(jit_uni_x8s8s32x_convolution_fwd_t<sse41>))
+        CPU_INSTANCE_AARCH64(jit_sve_512_x8s8s32x_convolution_fwd_t<s8, f32>)
+        REG_CONV_P_FWD(CPU_INSTANCE(_gemm_x8s8s32x_convolution_fwd_t<s8, f32>))
+        REG_CONV_P_FWD(CPU_INSTANCE(ref_convolution_int8_fwd_t))
+        REG_CONV_P_FWD(CPU_INSTANCE(ref_fused_convolution_fwd_t))
+        nullptr,
+    }},
+```
+
+Additionally, to overcome some of [2], instantiations of same classes for
```

Comment:
What if the classes are not templates, will the linker throw them away when we exclude them from the impl list? I assume this case is similar to what we already have for primitive kinds and propagation kinds.

Reply:
I left this paragraph here since `isa` is **always** a template argument. There's no way we split instantiations into separate translation units (it doesn't really make sense), that's why they will have to be wrapped as well to throw unneeded instantiations. If something is not a template, it has to be in a different translation unit to be thrown away by linker, which is the case for separation by propagation kind which we may want to split (I mentioned it earlier) for further tuning.

Reply:
So if we we specify AVX2 ISA then the following code will be thrown away at the link time? 
https://github.com/oneapi-src/oneDNN/blob/3431bd9e8092cdc87f36859ecf321e66545ee242/src/cpu/cpu_convolution_list.cpp#L133-L133

Reply:
Lists contain only calls to instances. These calls would be controlled by macros. They will appear/disappear based on user choice. But actual instantiations are living in a separate place, and they are the problem.
There are three possible scenarios:
1) A struct either non-templated or templated and enabled in the list. It will be included into final binary as expected.
2) A struct in non-templated and disabled in the list. It will be excluded from final binary as expected.
3) A struct is templated and disabled in the list. This is the situation I described in RFC. And it deserves some more attention:
    * Usually instantiations coming in a pack, like sse41, avx2, avx512_core, ...
    * If user enabled one of those, since all of them are living in the same translation unit, linker, by default, will pick them all.
    * That's why we need to wrap such instances with enabling macros as well to throw them manually in compile time from the final binary. Otherwise, we save nothing or almost nothing.

Does it make a situation clear? Thanks.

Reply:
I guess the 2nd bullet answers my question. Thanks.